### PR TITLE
feat(apollo-wind): add autoGrow in textArea [MST-12041]

### PR DESCRIPTION
**Pull request by ayush578**
## Summary

Adds opt-in **auto-grow** to the apollo-wind `Textarea` via new `minRows` / `maxRows` props (mirroring how `ApTextArea` in apollo-react exposes them, but implemented natively — apollo-wind has no MUI dependency).

### Behavior
- **Backward compatible:** with neither prop set, the component is byte-for-byte the same plain `<textarea>` as before.
- **Auto-grow:** floors at `minRows`, grows with content up to `maxRows`, then scrolls.
- **Uncapped manual resize:** the ceiling is enforced in JS (`height`), never via CSS `max-height`, so the bottom-right resize handle can be dragged past `maxRows`. A `ResizeObserver` detects the drag and hands sizing control to the user.
- Row heights are derived from the element's *computed* line-height + padding/border, so sizing is correct across `text-base`/`text-sm` and all themes.

### demo

https://github.com/user-attachments/assets/0c3b1957-6827-42f1-a439-c78c9bf464e8

Note: jsdom has no layout engine, so the tests cover the structural guarantees; the pixel-level grow/scroll is best confirmed in the new Storybook `AutoGrow` story.

### CODE REVIEW

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 1]
```diff
@@ -2,20 +2,102 @@ import * as React from 'react';
...
+
     return (
       <textarea
         className={cn(
           // Base styles (all themes)
           'flex min-h-[80px] w-full rounded-md border border-input bg-transparent px-3 py-2 text-base transition-colors placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm',
           // Future Dark / Future Light overrides
           'future:rounded-xl future:border-0 future:bg-surface-overlay future:text-sm future:placeholder:text-foreground-muted future:placeholder:font-normal future:focus-visible:ring-offset-2 future:focus-visible:ring-offset-background',
+          // Keep the resize handle available (and uncapped) when auto-growing.
+          autoGrow && 'resize-y',
```
[the remark this answers is not in the document]
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Fixed in 69bbc899. In auto-grow mode the base `min-h-[80px]` class is now dropped and the floor is owned by inline `min-height` (from `minRows`, or none). Since the rendered height is clamped to `maxHeight` in JS, `offsetHeight` can no longer exceed the ceiling and falsely trip `manuallyResized`. Auto-grow is also now gated on `maxRows` only.

File: packages/apollo-wind/src/components/ui/textarea.stories.tsx [hunk 2]
```diff
@@ -49,6 +49,24 @@ export const WithRows: Story = {
...
 
+export const AutoGrow = {
+  render: () => (
+    <div className="grid w-full max-w-sm gap-1.5">
+      <Label htmlFor="autogrow">Auto-growing field</Label>
+      <Textarea
+        id="autogrow"
+        minRows={3}
+        maxRows={5}
+        placeholder="Type multiple lines — the field grows from 3 rows up to 5, then scrolls. Drag the bottom-right handle to make it taller than 5 rows."
```
[the remark this answers is not in the document]
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.stories.tsx]: Fixed in 69bbc899 — em dashes removed from the story copy (placeholder now uses separate sentences).

File: packages/apollo-wind/src/components/ui/textarea.stories.tsx [hunk 3]
```diff
@@ -49,6 +49,24 @@ export const WithRows: Story = {
...
+      />
+      <p className="text-sm text-muted-foreground">
+        <code>minRows={3}</code> floor, auto-grows to <code>maxRows={5}</code>, then scrolls. The
+        resize handle is not capped — drag past 5 rows to take manual control.
+      </p>
```
[the remark this answers is not in the document]
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.stories.tsx:66]: Fixed in 69bbc899 — em dash removed from the description paragraph.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 4]
```diff
@@ -2,20 +2,113 @@ import * as React from 'react';
...
+    // hand sizing control to the user (stop auto-sizing) so the field stays put.
+    React.useLayoutEffect(() => {
+      const el = innerRef.current;
+      if (!el || maxRows === undefined || typeof ResizeObserver === 'undefined') return;
+
+      const observer = new ResizeObserver(() => {
+        const { lineHeight, chrome } = measureRowMetrics(el);
+        const maxHeight = maxRows * lineHeight + chrome;
+        if (el.offsetHeight > maxHeight + 1) manuallyResized.current = true;
+      });
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx:164]: ### 🟠 H1 — Width-driven reflow is not remeasured

The textarea is refit only when `value` or `defaultValue` changes. Its `ResizeObserver` fires when `w-full` or flex layouts change width, but this callback only checks height. Existing text can therefore wrap beyond the stale height while `overflow-y` remains hidden until the next keystroke; the `md:text-sm` metric change also leaves stale row calculations.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx:164]: Suggested change:

Track the observed width and rerun `resize()` when it changes, and remeasure row metrics when responsive or theme styles change, while keeping height-only user drags on the manual-resize path. Add a regression test that narrows the element after setting a wrapping value and asserts that height and overflow are recomputed.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 5]
```diff
@@ -2,20 +2,113 @@ import * as React from 'react';
...
+      [ref]
+    );
+
+    // Apply the `minRows` floor (as an inline min-height, which also bounds how far the
+    // resize handle can shrink) and, when `maxRows` is set, fit the height to the content
+    // up to the ceiling (scrolling past it). Once the user drags past the ceiling we stop
+    // auto-sizing so their height sticks.
+    const resize = React.useCallback(() => {
+      const el = innerRef.current;
+      if (!el || !managed) return;
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx:95]: ### 🟠 H2 — Disabling auto-grow leaves stale inline styles

When `minRows` and `maxRows` become unset, this early return leaves the previously managed `minHeight`, `height`, and `overflowY` on the DOM node. Switching from auto-grow to floor-only also retains the old fixed height and overflow state, so the textarea does not return to its declared mode.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx:95]: Suggested change:

Add transition cleanup for styles owned by auto-grow whenever the managed props change or become unset. Restore values supplied through the consumer `style` prop and reset `manuallyResized`; cover both `maxRows → undefined` and `maxRows → minRows-only` rerenders.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 6]
```diff
@@ -2,20 +2,113 @@ import * as React from 'react';
...
+    // A rendered height above the ceiling can only come from a manual drag; once seen,
+    // hand sizing control to the user (stop auto-sizing) so the field stays put.
+    React.useLayoutEffect(() => {
+      const el = innerRef.current;
+      if (!el || maxRows === undefined || typeof ResizeObserver === 'undefined') return;
+
+      const observer = new ResizeObserver(() => {
+        const { lineHeight, chrome } = measureRowMetrics(el);
+        const maxHeight = maxRows * lineHeight + chrome;
+        if (el.offsetHeight > maxHeight + 1) manuallyResized.current = true;
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx]: ### 🟠 H3 — Manual handoff can permanently suppress scrolling

This marks the field as manually resized but leaves the last auto-grow `overflowY` value in place, which is commonly `hidden`. Every later `resize()` then returns through the manual branch, so content that exceeds the manually selected height has no scrollbar.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Suggested change:

When handing sizing control to the user, restore the consumer-provided overflow value or set `overflowY` to `auto`. Reset that manual state when the sizing props change, and add coverage for typing beyond a manually enlarged field.

File: packages/apollo-wind/src/components/ui/textarea.test.tsx [hunk 7]
```diff
@@ -149,4 +149,58 @@ describe('Textarea', () => {
...
+      expect(textarea.style.minHeight).toMatch(/^\d+px$/);
+      // ...but there is no content-driven height (that only happens with maxRows).
+      expect(textarea.style.height).toBe('');
+    });
+
+    it('auto-grows and enables the resize handle once maxRows is set', () => {
+      render(<Textarea maxRows={5} aria-label="Message" />);
+      const textarea = screen.getByRole('textbox') as HTMLTextAreaElement;
+      expect(textarea).toHaveClass('resize-y');
+    });
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.test.tsx:166]: ### 🟠 H4 — The auto-grow test does not exercise auto-growth

Despite the test title, the only assertion checks the resize-handle class. None of the new tests verifies computed height, min/max clamping, overflow transitions, `ResizeObserver` handoff, prop removal, or width-driven reflow, so the core sizing contract remains untested.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.test.tsx:166]: Suggested change:

Mock `scrollHeight`, `offsetHeight`, computed row metrics, and `ResizeObserver` to assert growth, clamping, overflow, manual handoff, and prop cleanup. Add a browser-level component test for width-driven wrapping, which JSDOM cannot model.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 8]
```diff
@@ -2,20 +2,113 @@ import * as React from 'react';
...
+      el.style.minHeight = minRows !== undefined ? `${minHeight}px` : '';
+
+      // Floor-only (no `maxRows`): keep a fixed height, let content scroll. Nothing else to do.
+      if (maxRows === undefined || manuallyResized.current) return;
+
+      const maxHeight = maxRows * lineHeight + chrome;
+      el.style.height = 'auto';
+      const target = Math.min(Math.max(el.scrollHeight, minHeight), maxHeight);
+      el.style.height = `${target}px`;
+      el.style.overflowY = el.scrollHeight > maxHeight ? 'auto' : 'hidden';
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx]: ### 🟡 M1 — Border-box sizing clips by the border thickness

`maxHeight` and `minHeight` include the vertical border through `chrome`, while `scrollHeight` excludes borders. Assigning raw `scrollHeight` as a border-box height makes the content box short by the border thickness and can keep overflow hidden at the boundary.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Suggested change:

Expose the measured vertical border separately and use `scrollHeight + verticalBorder` as the target and overflow comparison in border-box mode. Keep a corresponding content-box calculation for consumer overrides, and add an exact-boundary test.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 9]
```diff
@@ -2,20 +2,113 @@ import * as React from 'react';
...
+      const { lineHeight, chrome } = measureRowMetrics(el);
+      const minHeight = minRows !== undefined ? minRows * lineHeight + chrome : 0;
+
+      // Own the floor explicitly; the base `min-h` class is dropped in managed mode.
+      el.style.minHeight = minRows !== undefined ? `${minHeight}px` : '';
+
+      // Floor-only (no `maxRows`): keep a fixed height, let content scroll. Nothing else to do.
+      if (maxRows === undefined || manuallyResized.current) return;
+
+      const maxHeight = maxRows * lineHeight + chrome;
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx]: ### 🟡 M2 — Conflicting row bounds disable auto-sizing

With `minRows > maxRows`, the inline `min-height` exceeds the JavaScript ceiling. CSS wins, then the observer interprets the oversized result as a manual drag and permanently disables auto-sizing.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Suggested change:

Normalize row counts to positive finite integers and compute `effectiveMinRows = Math.min(minRows, maxRows)` when both are supplied, matching the existing `prompt-editor` contract. Add a regression test for conflicting bounds.

File: packages/apollo-wind/src/components/ui/textarea.stories.tsx [hunk 10]
```diff
@@ -49,6 +49,25 @@ export const WithRows: Story = {
...
+      <Label htmlFor="autogrow">Auto-growing field</Label>
+      <Textarea
+        id="autogrow"
+        minRows={3}
+        maxRows={7}
+        placeholder="Starts at 3 rows and never shrinks below that. Grows with content up to 7 rows, then scrolls. Drag the bottom-right handle to make it taller than 7 rows."
+      />
+      <p className="text-sm text-muted-foreground">
+        <code>minRows={3}</code> is a floor: the field never renders or resizes shorter than 3 rows.
+        <code>maxRows={7}</code> auto-grows with content up to 7 rows, then scrolls. The resize
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.stories.tsx]: ### 🔵 L1 — The story drops the literal JSX braces

Inside JSX text, `<code>minRows={3}</code>` and `<code>maxRows={7}</code>` render as `minRows=3` and `maxRows=7`, so the documented prop syntax is not what users see.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.stories.tsx]: Suggested change:

Render the examples as string expressions: `<code>{'minRows={3}'}</code>` and `<code>{'maxRows={7}'}</code>`.

File: packages/apollo-wind/src/components/ui/textarea.tsx (hunk 4 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx:164]: Fixed in 0935eeb6. The ResizeObserver now tracks width and reruns `resize()` on width/breakpoint reflow (remeasuring row metrics), separate from the manual-drag path. Added a stubbed-layout test asserting a width change refits height + overflow.

File: packages/apollo-wind/src/components/ui/textarea.tsx (hunk 5 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx:95]: Fixed in 0935eeb6. Added a cleanup effect keyed on the sizing contract: managed→unmanaged restores the consumer's `height`/`minHeight`/`overflowY` (or clears ours), and auto-grow→floor-only drops the fitted height/overflow. `manuallyResized` is reset too. Tests cover both transitions.

File: packages/apollo-wind/src/components/ui/textarea.tsx (hunk 6 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Fixed in 0935eeb6. On manual handoff we now set `overflowY` to the consumer's value or `auto` (never frozen `hidden`), and the latch resets when the sizing props change. Added a test that types beyond a manually enlarged field and asserts overflow stays scrollable.

File: packages/apollo-wind/src/components/ui/textarea.test.tsx (hunk 7 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.test.tsx:166]: Fixed in 0935eeb6. New 'sizing behavior (stubbed layout)' suite mocks getComputedStyle metrics, `scrollHeight`/`offsetHeight`/`clientWidth` and `ResizeObserver`, asserting growth, min/max clamping, overflow transitions, border-box correction, inverted bounds, manual handoff, width reflow, and prop cleanup. Pixel-level text wrapping remains browser-only, but the width-reflow refit path is now covered via the observer.

File: packages/apollo-wind/src/components/ui/textarea.tsx (hunk 8 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Fixed in 0935eeb6. `measureRowMetrics` exposes the border separately; in border-box mode the content target and overflow comparison use `scrollHeight + verticalBorder`, with a content-box (`scrollHeight - padding`) branch for consumer overrides. Added a border-box height test.

File: packages/apollo-wind/src/components/ui/textarea.tsx (hunk 9 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.tsx]: Fixed in 0935eeb6. Rows are normalized to positive integers and `effMinRows = Math.min(minRows, maxRows)` when both are supplied, so the floor never exceeds the ceiling and the observer can't misread it as a manual drag. Added an inverted-bounds regression test.

File: packages/apollo-wind/src/components/ui/textarea.stories.tsx (hunk 10 above)
Comment (ayush578) [reply packages/apollo-wind/src/components/ui/textarea.stories.tsx]: Fixed in 0935eeb6 — rendered as `{'minRows={3}'}` / `{'maxRows={7}'}` so the literal braces display.

File: packages/apollo-wind/src/components/ui/textarea.tsx [hunk 11]
```diff
@@ -2,20 +2,189 @@ import * as React from 'react';
...
+      if (!el || !managed || typeof ResizeObserver === 'undefined') return;
+
+      let lastWidth = el.clientWidth;
+      const observer = new ResizeObserver(() => {
+        const width = el.clientWidth;
+        if (width !== lastWidth) {
+          lastWidth = width;
+          resize();
+          return;
+        }
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.tsx:154]: ### 🟠 H1 follow-up — Typography metric changes remain unobserved

Width-driven reflow is now handled, but this callback reruns `resize()` only when `clientWidth` changes. A fixed-width field crossing `md:text-sm` or changing theme/font metrics keeps the old pixel height and `overflow-y` state, so `minRows`/`maxRows` can remain stale and content can be clipped until the value changes.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.tsx:154]: Suggested change:

Trigger `resize()` on viewport breakpoint and relevant theme/style metric changes even when the element width is unchanged; compare the measured line height, padding, border, and box-sizing before deciding nothing changed. Add a regression test that keeps `clientWidth` constant while changing the computed line height and asserts that height and overflow are recomputed.

File: packages/apollo-wind/src/components/ui/textarea.test.tsx [hunk 12]
```diff
@@ -149,4 +149,28 @@ describe('Textarea', () => {
...
+
+    it('still forwards onChange while auto-growing', async () => {
+      const user = userEvent.setup();
+      const handleChange = vi.fn();
+      render(<Textarea minRows={3} maxRows={5} onChange={handleChange} aria-label="Message" />);
+      await user.type(screen.getByRole('textbox'), 'hi');
+      expect(handleChange).toHaveBeenCalled();
+    });
+  });
```
Comment (SreedharAvvari) [inline packages/apollo-wind/src/components/ui/textarea.test.tsx:175]: ### 🟠 H4 reopened — The sizing behavior suite was lost from the current head

This file still contains only structural class and `onChange` assertions. The stubbed-layout suite cited in the resolution existed in `0935eeb6` but is absent after the squash; the now-unused `act`, `beforeEach`, and `afterEach` imports are remnants of that lost block. Growth, clamping, overflow, manual handoff, reflow, bounds, and prop cleanup therefore remain untested.

Comment (SreedharAvvari) [reply packages/apollo-wind/src/components/ui/textarea.test.tsx:175]: Suggested change:

Restore the stubbed-layout suite from `0935eeb6` or equivalent coverage on the current branch, including exact assertions for border-box sizing, min/max clamping, manual overflow handoff, width and metric reflow, inverted bounds, and both prop-cleanup transitions. Keep the imports only when the restored tests use them.

**Review by SreedharAvvari — APPROVED** [review]
